File: react-app/src/wear/wearSync.ts
import { NativeAndroid } from "../native/nativeAndroid";
import { Preferences } from "../storage/preferences";

/**
 * Wear OS state sync. Pushes favourites, subscriptions, played ids, history and progress to
 * the companion watch app using the same Data Layer paths as the legacy Kotlin build, and
 * merges state received back from the watch.
 */
export function pushWearState(): void {
  if (!NativeAndroid.isAvailable()) return;
  try {
    const favourites = Preferences.getFavorites();
    const progressMap: Record<string, number> = {};
    for (const episodeId of Preferences.getPlayedEpisodeIds()) {
      progressMap[episodeId] = 0;
    }
    for (const entry of Preferences.getPodcastHistory()) {
      progressMap[entry.id] = Preferences.getEpisodeProgress(entry.id) * 1000;
    }
    const lastFm = Preferences.getLastFm();

    const payload = {
      favourite_ids: favourites,
      favourite_order: favourites,
      subscribed_podcast_ids: Preferences.getSubscribedPodcasts(),
      played_episode_ids: Preferences.getPlayedEpisodeIds(),
      history_episode_ids: Preferences.getPodcastHistory().map((entry) => entry.id),
      history_meta_json: JSON.stringify(Preferences.getPodcastHistory()),
      episode_progress_json: JSON.stringify(progressMap),
      lastfm_session_key: lastFm.sessionKey,
      lastfm_username: lastFm.username,
      lastfm_direct_enabled: lastFm.direct,
      lastfm_broadcast_enabled: lastFm.broadcast,
      lastfm_scrobble_podcasts: lastFm.podcasts
    };
    NativeAndroid.pushWearState(JSON.stringify(payload));
  } catch (error) {
    console.warn("Wear state push failed:", error);
  }
}

/** Merges state received from the watch into local preferences. */
export function applyWearState(rawPayload: string): void {
  try {
    const payload = JSON.parse(rawPayload);
    if (payload?.request) {
      pushWearState();
      return;
    }
    if (Array.isArray(payload.favourite_order) && payload.favourite_order.length > 0) {
      Preferences.setFavorites(payload.favourite_order);
    }
    if (payload.has_subscription_snapshot && Array.isArray(payload.subscribed_podcast_ids)) {
      Preferences.setSubscribedPodcasts(payload.subscribed_podcast_ids);
    }
    if (payload.has_episode_snapshot && Array.isArray(payload.played_episode_ids)) {
      for (const episodeId of payload.played_episode_ids) {
        if (!Preferences.isEpisodePlayed(episodeId)) Preferences.markEpisodePlayed(episodeId);
      }
      if (payload.episode_progress_json) {
        const progress = JSON.parse(payload.episode_progress_json);
        for (const [episodeId, positionMs] of Object.entries<number>(progress)) {
          if (positionMs > 0 && !Preferences.isEpisodePlayed(episodeId)) {
            Preferences.setEpisodeProgress(episodeId, Math.round(positionMs / 1000));
          }
        }
      }
      if (payload.history_meta_json) {
        const history = JSON.parse(payload.history_meta_json);
        if (Array.isArray(history)) {
          for (const entry of history) {
            if (entry?.id) Preferences.addPodcastHistory(entry);
          }
        }
      }
    }
  } catch (error) {
    console.warn("Wear state merge failed:", error);
  }
}

let started = false;

/** Starts listening for state received from the watch and pushes the initial snapshot. */
export function initWearSync(): void {
  if (started) return;
  started = true;
  try {
    NativeAndroid.addWearStateListener(applyWearState);
    pushWearState();
  } catch {
    // Wear support is optional.
  }
}

File: react-app/modules/native-android/android/src/main/java/com/hyliankid14/bbcradioplayer/nativeandroid/WearSync.kt
package com.hyliankid14.bbcradioplayer.nativeandroid

import android.content.Context
import com.google.android.gms.wearable.DataMapItem
import com.google.android.gms.wearable.PutDataMapRequest
import com.google.android.gms.wearable.Wearable
import org.json.JSONObject

/**
 * Pushes phone state to the Wear OS companion using the same Data Layer paths and keys as the
 * legacy Kotlin `WearAppStateSync`, so the existing watch app keeps working unchanged.
 */
object WearSync {

  const val PATH_APP_STATE = "/bbcradioplayer/state"
  const val PATH_REQUEST_STATE = "/bbcradioplayer/request_state"
  const val PATH_STATE_PAYLOAD = "/bbcradioplayer/state_payload"

  private const val KEY_FAVOURITE_IDS = "favourite_ids"
  private const val KEY_FAVOURITE_ORDER = "favourite_order"
  private const val KEY_SUBSCRIBED_PODCAST_IDS = "subscribed_podcast_ids"
  private const val KEY_HAS_SUBSCRIPTION_SNAPSHOT = "has_subscription_snapshot"
  private const val KEY_PLAYED_EPISODE_IDS = "played_episode_ids"
  private const val KEY_HISTORY_EPISODE_IDS = "history_episode_ids"
  private const val KEY_HISTORY_META_JSON = "history_meta_json"
  private const val KEY_EPISODE_PROGRESS_JSON = "episode_progress_json"
  private const val KEY_HAS_EPISODE_SNAPSHOT = "has_episode_snapshot"
  private const val KEY_UPDATED_AT = "updated_at"
  private const val KEY_LASTFM_SESSION_KEY = "lastfm_session_key"
  private const val KEY_LASTFM_USERNAME = "lastfm_username"
  private const val KEY_LASTFM_DIRECT_ENABLED = "lastfm_direct_enabled"
  private const val KEY_LASTFM_BROADCAST_ENABLED = "lastfm_broadcast_enabled"
  private const val KEY_LASTFM_SCROBBLE_PODCASTS = "lastfm_scrobble_podcasts"

  /**
   * Pushes the given state (built by the JS layer) to the watch. [payloadJson] contains the
   * favourites, subscriptions, played ids, history, progress and Last.fm fields.
   */
  fun pushState(context: Context, payloadJson: String) {
    val payload = try {
      JSONObject(payloadJson)
    } catch (_: Exception) {
      return
    }

    val request = PutDataMapRequest.create(PATH_APP_STATE).apply {
      dataMap.putStringArrayList(KEY_FAVOURITE_IDS, payload.stringList(KEY_FAVOURITE_IDS))
      dataMap.putStringArrayList(KEY_FAVOURITE_ORDER, payload.stringList(KEY_FAVOURITE_ORDER))
      dataMap.putStringArrayList(KEY_SUBSCRIBED_PODCAST_IDS, payload.stringList(KEY_SUBSCRIBED_PODCAST_IDS))
      dataMap.putBoolean(KEY_HAS_SUBSCRIPTION_SNAPSHOT, true)
      dataMap.putStringArrayList(KEY_PLAYED_EPISODE_IDS, payload.stringList(KEY_PLAYED_EPISODE_IDS))
      dataMap.putStringArrayList(KEY_HISTORY_EPISODE_IDS, payload.stringList(KEY_HISTORY_EPISODE_IDS))
      dataMap.putBoolean(KEY_HAS_EPISODE_SNAPSHOT, true)
      dataMap.putString(KEY_EPISODE_PROGRESS_JSON, payload.optString(KEY_EPISODE_PROGRESS_JSON, "{}"))
      dataMap.putString(KEY_HISTORY_META_JSON, payload.optString(KEY_HISTORY_META_JSON, ""))
      dataMap.putString(KEY_LASTFM_SESSION_KEY, payload.optString(KEY_LASTFM_SESSION_KEY, ""))
      dataMap.putString(KEY_LASTFM_USERNAME, payload.optString(KEY_LASTFM_USERNAME, ""))
      dataMap.putBoolean(KEY_LASTFM_DIRECT_ENABLED, payload.optBoolean(KEY_LASTFM_DIRECT_ENABLED, false))
      dataMap.putBoolean(KEY_LASTFM_BROADCAST_ENABLED, payload.optBoolean(KEY_LASTFM_BROADCAST_ENABLED, true))
      dataMap.putBoolean(KEY_LASTFM_SCROBBLE_PODCASTS, payload.optBoolean(KEY_LASTFM_SCROBBLE_PODCASTS, false))
      dataMap.putLong(KEY_UPDATED_AT, System.currentTimeMillis())
    }.asPutDataRequest().setUrgent()

    try {
      Wearable.getDataClient(context).putDataItem(request)
    } catch (_: Exception) {
    }
  }

  /** Converts a received watch data item into a JSON string for the JS layer to merge. */
  fun readIncomingState(item: DataMapItem): String {
    val dataMap = item.dataMap
    val progressJson = dataMap.getString(KEY_EPISODE_PROGRESS_JSON).orEmpty()
    return JSONObject().apply {
      put(KEY_FAVOURITE_IDS, org.json.JSONArray(dataMap.getStringArrayList(KEY_FAVOURITE_IDS) ?: emptyList<String>()))
      put(KEY_FAVOURITE_ORDER, org.json.JSONArray(dataMap.getStringArrayList(KEY_FAVOURITE_ORDER) ?: emptyList<String>()))
      put(KEY_HAS_SUBSCRIPTION_SNAPSHOT, dataMap.getBoolean(KEY_HAS_SUBSCRIPTION_SNAPSHOT, true))
      put(KEY_SUBSCRIBED_PODCAST_IDS, org.json.JSONArray(dataMap.getStringArrayList(KEY_SUBSCRIBED_PODCAST_IDS) ?: emptyList<String>()))
      put(KEY_HAS_EPISODE_SNAPSHOT, dataMap.getBoolean(KEY_HAS_EPISODE_SNAPSHOT, true))
      put(KEY_PLAYED_EPISODE_IDS, org.json.JSONArray(dataMap.getStringArrayList(KEY_PLAYED_EPISODE_IDS) ?: emptyList<String>()))
      put(KEY_EPISODE_PROGRESS_JSON, progressJson)
      put(KEY_HISTORY_META_JSON, dataMap.getString(KEY_HISTORY_META_JSON).orEmpty())
      put(KEY_LASTFM_SESSION_KEY, dataMap.getString(KEY_LASTFM_SESSION_KEY).orEmpty())
      put(KEY_LASTFM_USERNAME, dataMap.getString(KEY_LASTFM_USERNAME).orEmpty())
      put(KEY_LASTFM_DIRECT_ENABLED, dataMap.getBoolean(KEY_LASTFM_DIRECT_ENABLED, false))
      put(KEY_LASTFM_BROADCAST_ENABLED, dataMap.getBoolean(KEY_LASTFM_BROADCAST_ENABLED, true))
      put(KEY_LASTFM_SCROBBLE_PODCASTS, dataMap.getBoolean(KEY_LASTFM_SCROBBLE_PODCASTS, false))
    }.toString()
  }

  private fun JSONObject.stringList(key: String): ArrayList<String> {
    val array = optJSONArray(key) ?: return ArrayList()
    val list = ArrayList<String>(array.length())
    for (i in 0 until array.length()) {
      val value = array.optString(i, "")
      if (value.isNotBlank()) list.add(value)
    }
    return list
  }
}
